### PR TITLE
aks-preview: add open-service-mesh as an aks addon

### PR DESCRIPTION
**Pull request by shashankram**
This change extends the `az aks` cli commands to allow
enabling/disabling the open-service-mesh AKS addon as a part
of `az aks create` and `az aks enable-addons` and
`az aks disable-addons` commands.

Currently an optional mesh name can be specified while
enabling the addon.

Signed-off-by: Shashank Ram <shashank08@gmail.com>

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [X] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [X] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

File: src/aks-preview/azext_aks_preview/_params.py [hunk 1]
```diff
@@ -180,6 +180,7 @@ def load_arguments(self, _):
         c.argument('addons', options_list=['--addons', '-a'], validator=validate_addons)
         c.argument('subnet_name', options_list=['--subnet-name', '-s'])
         c.argument('disable_sgxquotehelper', action='store_true')
+        c.argument('osm_mesh_name', options_list=['--osm-mesh-name'])
```
Comment (robbiezhang) [inline src/aks-preview/azext_aks_preview/_params.py:183]: although we validate the mesh-name in the server side, it's better to have the same validation on the client side.

File: src/aks-preview/azext_aks_preview/custom.py [hunk 2]
```diff
@@ -2447,13 +2466,14 @@ def aks_disable_addons(cmd, client, resource_group_name, name, addons, no_wait=F
...
-                      subnet_name=None, appgw_name=None, appgw_subnet_prefix=None, appgw_id=None, appgw_subnet_id=None, appgw_watch_namespace=None, disable_sgxquotehelper=False, no_wait=False):
+                      subnet_name=None, appgw_name=None, appgw_subnet_prefix=None, appgw_id=None, appgw_subnet_id=None,
+                      appgw_watch_namespace=None, disable_sgxquotehelper=False, no_wait=False, osm_mesh_name=None):
     instance = client.get(resource_group_name, name)
     subscription_id = get_subscription_id(cmd.cli_ctx)
     instance = _update_addons(cmd, instance, subscription_id, resource_group_name, name, addons, enable=True,
                               workspace_resource_id=workspace_resource_id, subnet_name=subnet_name,
                               appgw_name=appgw_name, appgw_subnet_prefix=appgw_subnet_prefix, appgw_id=appgw_id, appgw_subnet_id=appgw_subnet_id, appgw_watch_namespace=appgw_watch_namespace,
-                              disable_sgxquotehelper=disable_sgxquotehelper, no_wait=no_wait)
+                              disable_sgxquotehelper=disable_sgxquotehelper, no_wait=no_wait, osm_mesh_name=osm_mesh_name)
```
Comment (robbiezhang) [inline src/aks-preview/azext_aks_preview/custom.py:2476]: nit: move it before the no_wait

Comment (shashankram) [reply src/aks-preview/azext_aks_preview/custom.py]: Done.

File: src/aks-preview/azext_aks_preview/_params.py (hunk 1 above)
Comment (shashankram) [reply src/aks-preview/azext_aks_preview/_params.py:183]: Considering this argument is optional and just a string, does backend validation not suffice? This property gets applied as a value for a kubernetes label.

Comment (robbiezhang) [reply src/aks-preview/azext_aks_preview/_params.py:183]: It is used in the label and webhook configuration name. They all need to satisfy the DNS naming validation

**Comment by yonzhan** [comment]
aks

File: src/aks-preview/azext_aks_preview/_params.py (hunk 1 above)
Comment (shashankram) [reply src/aks-preview/azext_aks_preview/_params.py:183]: Yes, but doesn't backend validation suffice?

Comment (shashankram) [reply src/aks-preview/azext_aks_preview/_params.py:183]: @robbiezhang we discussed and this should be ok as is at the moment.

**Review by robbiezhang — APPROVED** [review]

File: src/aks-preview/azext_aks_preview/_consts.py [hunk 3]
```diff
@@ -25,6 +25,10 @@
 CONST_INGRESS_APPGW_SUBNET_PREFIX = "subnetPrefix"
 CONST_INGRESS_APPGW_WATCH_NAMESPACE = "watchNamespace"
 
+# Open Service Mesh configuration keys
+CONST_OSM_ADDON_NAME = "openServiceMesh"
```
Comment (gtracer) [inline src/aks-preview/azext_aks_preview/_consts.py:29]: I would prefer we, like the commend above, use a full name here. So instead of OSM call it OPEN_SERVICE_MESH. This is doubly true since we have another addon called OMS already and I think it would be easy to mistake the two if you are reading quickly

File: src/aks-preview/azext_aks_preview/custom.py [hunk 4]
```diff
@@ -851,7 +852,8 @@ def aks_create(cmd,     # pylint: disable=too-many-locals,too-many-statements,to
                aad_admin_group_object_ids=None,
                disable_sgxquotehelper=False,
                assign_identity=None,
-               no_wait=False):
+               no_wait=False,
+               osm_mesh_name=None):
```
Comment (gtracer) [inline src/aks-preview/azext_aks_preview/custom.py:856]: Nit: move this up to be with the other addons or at least above no_wait

File: src/aks-preview/azext_aks_preview/_params.py (hunk 1 above)
Comment (gtracer) [inline src/aks-preview/azext_aks_preview/_params.py:183]: Is there any validation we can do on the input param? For example, on name length or allowed chars etc?

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py [hunk 5]
```diff
@@ -270,6 +270,112 @@ def test_aks_byo_appgw_with_ingress_appgw_addon(self, resource_group, resource_g
             'addon_client_id': addon_client_id,
         })
 
+    @AllowLargeResponse()
+    @ResourceGroupPreparer(random_name_length=17, name_prefix='clitest', location='westus2')
+    def test_aks_create_with_openservicemesh_addon(self, resource_group, resource_group_location):
```
Comment (gtracer) [inline src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:275]: Thank you for adding tests

File: src/aks-preview/azext_aks_preview/_consts.py (hunk 3 above)
Comment (shashankram) [reply src/aks-preview/azext_aks_preview/_consts.py]: Done

File: src/aks-preview/azext_aks_preview/custom.py (hunk 4 above)
Comment (shashankram) [reply src/aks-preview/azext_aks_preview/custom.py]: Done.

File: src/aks-preview/azext_aks_preview/_params.py (hunk 1 above)
Comment (shashankram) [reply src/aks-preview/azext_aks_preview/_params.py:183]: Discussed this with Robbie, for a start we prefer to keep the validations in the backend. We are still discussing some design considerations and might even consider removing this option in the near term.

Comment (gtracer) [reply src/aks-preview/azext_aks_preview/_params.py:183]: Ok, if you keep it then please add the validator before GA

**Review by gtracer — APPROVED** [review]

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py [hunk 6]
```diff
@@ -270,6 +270,112 @@ def test_aks_byo_appgw_with_ingress_appgw_addon(self, resource_group, resource_g
...
+        self.cmd(create_cmd, checks=[
+            self.check('provisioningState', 'Succeeded'),
+            self.check('addonProfiles.openServiceMesh.enabled', True),
+        ])
+
+    @AllowLargeResponse()
+    @ResourceGroupPreparer(random_name_length=17, name_prefix='clitest', location='westus2')
+    def test_aks_create_with_openservicemesh_addon_with_mesh_name(self, resource_group, resource_group_location):
+        aks_name = self.create_random_name('cliakstest', 16)
+        osm_mesh_name = self.create_random_name('cliakstest', 16)
```
Comment (zqingqing1) [inline src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:293]: nit: name another osm_mesh_name? cliakstest seems not related to the mesh name.

**Review by zqingqing1 — APPROVED** [review]

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py (hunk 6 above)
Comment (shashankram) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:293]: I see what you mean, but any string that meets k8s label value requirements works fine here. I'll consider chating this to ue a deterministic name in a follow-up.

**Review by arrownj — APPROVED** [review]
LGTM
